src/experimental/queue.rs

```diff
@@ -0,0 +1,372 @@
+                        }
+                        if conns.contains_key(&req.db_name[..]) {
+                            req.response_sender.send("ok".into()).await;
+                            continue;
+                        }
+
+                        // Create a new Connection and connection_loop.
+                        let (query_req_ch_snd, query_req_ch_rec) = channel::<Request>(1);
+                        let (query_quit_ch_snd, query_quit_ch_rec) = channel::<()>(1);
+                        let c = Connection {
+                            query_ch: query_req_ch_snd,
+                            quit_ch: query_quit_ch_snd,
+                        };
+                        conns.insert(req.db_name.clone(), c);
+                        // TODO do something real to open the db.
+                        let store = Store {
+                            map: Box::new(HashMap::new()),
+                        };
+                        spawn_local(connection_loop(query_req_ch_rec, query_quit_ch_rec, store));
+                        req.response_sender.send("ok".into()).await;
+                        continue;
+                    }
+                    "close" => {
+                        let msg = if req.db_name.is_empty() {
+                            "db_name must be non-empty"
+                        } else {
+                            match conns.get(&req.db_name[..]) {
+                                Some(c) => {
+                                    c.quit_ch.send(()).await;
+                                    // TODO maybe wait for the quit to complete?
+                                    conns.remove(&req.db_name[..]);
+                                    "ok"
+                                }
+                                None => "db not open",
+                            }
+                        };
+                        req.response_sender.send(msg.into()).await;
+                        continue;
+                    }
+                    _ => match conns.get(&req.db_name[..]) {
+                        Some(c) => {
+                            c.query_ch.send(req).await;
+                        }
+                        None => {
+                            req.response_sender.send("no such db".into()).await;
+                        }
+                    },
+                }
+            }
+        }
+    }
+}
+
+// One per connection / database. Handles requests for this db.
+async fn connection_loop(req_ch_rec: Receiver<Request>, quit_ch_rec: Receiver<()>, mut store: Store) {
+    // We buffer Requests to the db so we can allow Requests to the current
+    // write tx (if any) to jump the queue.
+    let mut queue = VecDeque::new();
+
+    // If write_tx is None we are executing reads.
+    let mut write_tx: Option<WriteTx> = None;
+    let mut next_tx_id = 0u64;
+
+    loop {
+        // Check if we have received the quit signal.
+        match quit_ch_rec.try_recv() {
```

Comment:
I think there is probably a lighter-weight way to shut these things down than the go-style way.